bitcoin/test/fuzz/miniscript_random.cpp

```diff
@@ -241,93 +247,423 @@ std::optional<uint32_t> ConsumeTimeLock(FuzzedDataProvider& provider) {
+        case Fragment::WRAP_V:
+        case Fragment::WRAP_J:
+        case Fragment::WRAP_N:
+        case Fragment::AND_V:
+        case Fragment::AND_B:
+        case Fragment::OR_B:
+        case Fragment::OR_C:
+        case Fragment::OR_D:
+        case Fragment::OR_I:
+        case Fragment::ANDOR:
+            return {{subt, frag}};
+        case Fragment::THRESH: {
+            uint32_t children;
+            if (subt.size() < 2) {
+                children = subt.size();
+            } else {
+                // If we hit a thresh with 2 subnodes, artificially extend it to any number
+                // (2 or larger) by replicating the type of the last subnode.
+                children = provider.ConsumeIntegralInRange<uint32_t>(2, MAX_OPS_PER_SCRIPT / 2);
+            }
+            auto k = provider.ConsumeIntegralInRange<uint32_t>(1, children);
+            std::vector<Type> subs = subt;
+            while (subs.size() < children) subs.push_back(subs.back());
+            return {{std::move(subs), frag, k}};
+        }
+    }
+}
+
 /**
  * Generate a Miniscript node based on the fuzzer's input.
+ *
+ * - ConsumeNode is a function object taking a Type, and returning an std::optional<NodeInfo>.
+ * - root_type is the required type properties of the constructed NodeRef.
+ * - strict_valid sets whether ConsumeNode is expected to guarantee a NodeInfo that results in
+ *   a NodeRef whose Type() matches the type fed to ConsumeNode.
  */
-NodeRef GenNode(FuzzedDataProvider& provider) {
+template<typename F>
+NodeRef GenNode(F ConsumeNode, Type root_type = ""_mst, bool strict_valid = false) {
     /** A stack of miniscript Nodes being built up. */
     std::vector<NodeRef> stack;
     /** The queue of instructions. */
-    std::vector<std::optional<NodeInfo>> todo{{}};
+    std::vector<std::pair<Type, std::optional<NodeInfo>>> todo{{root_type, {}}};
 
     while (!todo.empty()) {
         // The expected type we have to construct.
-        if (!todo.back()) {
+        auto type_needed = todo.back().first;
+        if (!todo.back().second) {
             // Fragment/children have not been decided yet. Decide them.
-            auto node_info = ConsumeNode(provider);
-            uint8_t n_subs = node_info->n_subs;
+            auto node_info = ConsumeNode(type_needed);
             if (!node_info) return {};
-            todo.back() = std::move(node_info);
-            for (uint8_t i = 0; i < n_subs; i++) todo.push_back({});
+            auto subtypes = std::move(node_info)->subtypes;
+            todo.back().second = std::move(node_info);
+            todo.reserve(todo.size() + subtypes.size());
+            // As elements on the todo stack are processed back to front, construct
+            // them in reverse order (so that the first subnode is generated first).
+            for (size_t i = 0; i < subtypes.size(); ++i) {
+                todo.emplace_back(*(subtypes.rbegin() + i), std::nullopt);
+            }
```

Comment:
nit: Seems simpler to just keep the loop in order here
```cpp
for (auto type : subtypes) todo.emplace_back(type, std::nullopt);
```
And instead iterate back to front when popping the childs off the stack below?
```cpp
            // Gather children from the back of stack. Since they are added back to front to the stack, iterate the same way here.
            std::vector<NodeRef> sub;
            sub.reserve(info.n_subs);
            for (size_t i = 0; i < info.n_subs; ++i) {
                sub.push_back(std::move(*(stack.end() - 1 - i)));
            }
```

Anyways not a big deal, and good catch it completely slipped my review when we were working on the "not smart" version.

Reply:
I considered this, but didn't as that'd mean the data in the fuzzer seed for the last subexpression would come first. Not that it really matters, but I found it nicer to do everything left to right.